File: gui/plugins/artifact_view.py
#!/usr/bin/env python
"""This plugin adds artifact functionality to the UI."""

import itertools
import StringIO

from grr.gui import renderers
from grr.gui.plugins import fileview
from grr.gui.plugins import forms
from grr.gui.plugins import semantic
from grr.lib import aff4
from grr.lib import artifact
from grr.lib import artifact_lib
from grr.lib import parsers
from grr.lib import rdfvalue


class ArtifactListRenderer(forms.MultiSelectListRenderer):
  """Renderer for listing the available Artifacts."""

  type = rdfvalue.ArtifactName

  artifact_template = ("""
          <div id='{{unique|escape}}_artifact_description'>
            <h4><div name='artifact_name'/></h4>
            <div name='artifact_description'/>
            <table>
              <tr><td>Labels<td><div name='artifact_labels'/></tr>
              <tr><td>Platforms<td><div name='artifact_supported_os'/></tr>
              <tr><td>Conditions<td><div name='artifact_conditions'/></tr>
              <tr><td>Dependencies<td><div name='artifact_dependencies'/></tr>
              <tr><td>Links<td><div name='artifact_links'/></tr>
              <tr><td>Output Type<td><div name='artifact_output_type'/></tr>
            </table>
            <h5>Artifact Collectors</h5>
            <table name='artifact_collectors'>
              <tbody></tbody>
            </table>
            <h5>Artifact Processors</h5>
            <table name='artifact_processors'>
              <tbody></tbody>
            </table>
          </div>""")

  layout_template = (
      """<div class="control-group">"""
      + forms.TypeDescriptorFormRenderer.default_description_view + """
  <div id='{{unique|escape}}_artifact_renderer' class="controls">
  <div>
    <table class='artifact_table'>
      <tr>
        <td>
          <input id='{{unique|escape}}_search'
            placeholder="Search"></input><br>
          <select id='{{unique|escape}}_os_filter'
            placeholder="OS Filter"></input>
        <td>
        <td>
      </tr>
      <tr>
        <td class="artifact_table">
          <select id='{{unique|escape}}_artifact_list' class='artifact_list'
            multiple />
        <td class="artifact_table">
          <select id='{{this.prefix|escape}}' class='artifact_list' multiple/>
        <td class="artifact_table">"""
      + artifact_template + """
      </tr>
      <tr>
        <td>
          <a id='{{unique|escape}}_artifact_add'>Add</a>
          <a id='{{unique|escape}}_artifact_add_all'
            class='pull-right'>Add all  </a>
        <td>
          <a id='{{unique|escape}}_select_clear'>Clear</a>
          <a id='{{unique|escape}}_select_remove'
            class='pull-right'>Remove  </a>
        <td>
      </tr>
    </table>
  </div>
  </div>

</div>
""")

  def Layout(self, request, response):
    """Get available artifact information for display."""
    # Get all artifacts that aren't Bootstrap and aren't the base class.
    self.artifacts = {}
    artifact.LoadArtifactsFromDatastore(token=request.token)
    for name, artifact_val in artifact_lib.ArtifactRegistry.artifacts.items():
      if set(["Bootstrap"]).isdisjoint(artifact_val.labels):
        self.artifacts[name] = artifact_val
    self.labels = artifact_lib.ARTIFACT_LABELS

    # Convert artifacts into a dict usable from javascript.
    artifact_dict = {}
    for artifact_name, artifact_val in self.artifacts.items():
      artifact_dict[artifact_name] = artifact_val.ToExtendedDict()
      processors = []
      for processor in parsers.Parser.GetClassesByArtifact(artifact_name):
        processors.append({"name": processor.__name__,
                           "output_types": processor.output_types,
                           "doc": processor.GetDescription()})
      artifact_dict[artifact_name]["processors"] = processors

    # Skip the our parent and call the TypeDescriptorFormRenderer direct.
    response = renderers.TypeDescriptorFormRenderer.Layout(self, request,
                                                           response)
    return self.CallJavascript(response, "ArtifactListRenderer.Layout",
                               prefix=self.prefix,
                               artifacts=artifact_dict,
                               supported_os=artifact_lib.SUPPORTED_OS_LIST,
                               labels=self.labels)


class ArtifactRDFValueRenderer(semantic.RDFValueRenderer):
  """A special renderer for ArtifactRDFValues."""

  classname = "Artifact"

  layout_template = renderers.Template(
      """
<div id={{unique|escape}}_artifact_description>"""
      + ArtifactListRenderer.artifact_template + """
</div>
""")

  def Layout(self, request, response):
    self.artifact_str = self.proxy.ToPrettyJson()
    response = super(ArtifactRDFValueRenderer, self).Layout(request, response)
    return self.CallJavascript(response, "ArtifactRDFValueRenderer.Layout",
                               artifact_str=self.artifact_str)


class ArtifactRawRDFValueRenderer(semantic.RDFValueRenderer):
  """A renderer for showing JSON format for ArtifactRDFValues."""

  classname = "Artifact"

  layout_template = renderers.Template(
      "<pre>{{this.artifact_str|escape}}</pre>")

  def Layout(self, request, response):
    self.artifact_str = self.proxy.ToPrettyJson(extended=True)
    super(ArtifactRawRDFValueRenderer, self).Layout(request, response)


class ArtifactManagerView(renderers.TableRenderer):
  """Artifact Manager table with toolbar."""

  description = "Artifact Manager"
  behaviours = frozenset(["Configuration"])
  order = 50

  toolbar = "ArtifactManagerToolbar"

  def __init__(self, **kwargs):
    super(ArtifactManagerView, self).__init__(**kwargs)
    self.AddColumn(semantic.RDFValueColumn("Artifact Name", width="5%"))
    self.AddColumn(semantic.RDFValueColumn(
        "Artifact Details", width="50%", renderer=ArtifactRDFValueRenderer))
    self.AddColumn(semantic.RDFValueColumn(
        "Artifact Raw", width="40%", renderer=ArtifactRawRDFValueRenderer))

  def BuildTable(self, start_row, end_row, request):
    """Builds table artifacts."""
    artifact_urn = rdfvalue.RDFURN("aff4:/artifact_store")
    try:
      collection = aff4.FACTORY.Open(artifact_urn,
                                     aff4_type="RDFValueCollection",
                                     token=request.token)
    except IOError:
      return

    self.size = len(collection)
    row_index = start_row
    for value in itertools.islice(collection, start_row, end_row):
      self.AddCell(row_index, "Artifact Name", value.name)
      self.AddCell(row_index, "Artifact Details", value)
      self.AddCell(row_index, "Artifact Raw", value)
      row_index += 1

  def Layout(self, request, response):
    """Populate the table state with the request."""
    if self.toolbar:
      tb_cls = renderers.Renderer.classes[self.toolbar]
      tb_cls().Layout(request, response)
    return super(ArtifactManagerView, self).Layout(request, response)


class ArtifactManagerToolbar(renderers.TemplateRenderer):
  """A navigation enhancing toolbar.

  Internal State:
    - aff4_path: The path we are viewing now in the table.
  """
  post_parameters = ["aff4_path"]
  event_queue = "file_select"

  layout_template = renderers.Template("""
<ul id="toolbar_{{unique|escape}}" class="breadcrumb">
  <li>
    <button id='{{unique|escape}}_upload' class="btn"
      title="Upload Artifacts as JSON or YAML"
      data-toggle="modal" data-target="#upload_dialog_{{unique|escape}}">
      <img src='/static/images/upload.png' class='toolbar_icon'>
    </button>
  </li>
  <li>
    <button id='{{unique|escape}}_deleteall' class="btn"
      title="Delete all uploaded artifacts" data-toggle="modal"
      data-target="#delete_confirm_dialog_{{unique|escape}}">
      <img src='/static/images/editdelete.png' class='toolbar_icon'>
    </button>
  </li>

</ul>

<div id="upload_dialog_{{unique|escape}}" class="modal hide" tabindex="-1"
  role="dialog" aria-hidden="true">
  <div class="modal-header">
    <button id="upload_artifact_btn_{{unique|escape}}" type="button"
    class="close" data-dismiss="modal" aria-hidden="true">
      x</button>
    <h3>Upload File</h3>
  </div>
  <div class="modal-body" id="upload_dialog_body_{{unique|escape}}"></div>
  <div class="modal-footer">
    <button id="upload_artifact_close_btn_{{unique|escape}}" class="btn"
    data-dismiss="modal" aria-hidden="true">Close</button>
  </div>
</div>

<div id="delete_confirm_dialog_{{unique|escape}}"
  class="modal hide" tabindex="-1" role="dialog" aria-hidden="true">
</div>

""")

  def Layout(self, request, response):
    response = super(ArtifactManagerToolbar, self).Layout(request, response)
    return self.CallJavascript(response, "ArtifactManagerToolbar.Layout")


class DeleteArtifactsConfirmationDialog(renderers.ConfirmationDialogRenderer):
  """Dialog that asks for confirmation to delete uploaded artifacts.

  Note that this only deletes artifacts that have been uploaded via the
  ArtifactManager.  Artifacts loaded from the artifacts directory are
  unaffected.
  """

  content_template = renderers.Template("""
<p>Are you sure you want to <strong>delete all</strong>
uploaded artifacts?</p>
""")

  ajax_template = renderers.Template("""
<p class="text-info">Uploaded artifacts were deleted successfully.</p>
""")

  def RenderAjax(self, request, response):
    aff4.FACTORY.Delete("aff4:/artifact_store", token=request.token)
    return self.RenderFromTemplate(self.ajax_template, response,
                                   unique=self.unique, this=self)


class ArtifactJsonUploadView(fileview.UploadView):
  """Renders a binary upload page."""
  post_parameters = []
  upload_handler = "ArtifactUploadHandler"
  storage_path = "aff4:/artifact_store"


class ArtifactUploadHandler(fileview.UploadHandler):
  """Handles upload of a binary config file such as a driver."""

  def RenderAjax(self, request, response):
    """Handle the upload via ajax."""
    try:
      self.uploaded_file = request.FILES.items()[0][1]
      content = StringIO.StringIO()
      for chunk in self.uploaded_file.chunks():
        content.write(chunk)
      self.dest_path = artifact.UploadArtifactYamlFile(
          content.getvalue(), token=request.token)

      return renderers.TemplateRenderer.Layout(self, request, response,
                                               self.success_template)
    except (IOError, artifact_lib.ArtifactDefinitionError) as e:
      self.error = "Could not write artifact to database %s" % e
    return renderers.TemplateRenderer.Layout(self, request, response,
                                             self.error_template)

File: gui/static/javascript/plugins/artifact_view.js
var grr = window.grr || {};

/**
 * Namespace for artifacts.
 */
grr.artifact_view = {};


grr.Renderer('ArtifactRDFValueRenderer', {
  Layout: function(state) {
    var unique = state.unique;
    var artifact_str = state.artifact_str;

    var description_element = unique + '_artifact_description';
    var artifact_obj = JSON.parse(artifact_str);
    grr.artifact_view.renderArtifactFromObject(artifact_obj,
                                               description_element);
    // Remove heading to clean up display.
    $('div[name=artifact_name]').hide();
  }
});


grr.Renderer('ArtifactManagerToolbar', {
  Layout: function(state) {
    var unique = state.unique;

    $('#upload_dialog_' + unique).on('show.bs.modal', function() {
      grr.layout('ArtifactJsonUploadView',
                 'upload_dialog_body_' + unique);
    });

    $('#delete_confirm_dialog_' + unique).on('show.bs.modal', function() {
      grr.layout('DeleteArtifactsConfirmationDialog',
                 'delete_confirm_dialog_' + unique);
    });
  }
});


grr.Renderer('ArtifactListRenderer', {
  Layout: function(state) {
    // Populate the artifact manager.
    var selected_list_element = state.prefix;
    var artifact_list_element = state.unique + '_artifact_list';
    var os_filter_element = state.unique + '_os_filter';
    var search_element = state.unique + '_search';
    var description_element = state.unique + '_artifact_description';
    var artifact_div = $('#' + state.unique + '_artifact_renderer');

    var artifact_data = artifact_div.data();
    artifact_data.selected_list_element = selected_list_element;
    artifact_data.artifact_list_element = artifact_list_element;
    artifact_data.description_element = description_element;
    artifact_data.os_filter_element = os_filter_element;
    artifact_data.search_element = search_element;
    artifact_data.artifacts = state.artifacts;
    artifact_data.labels = state.labels;

    // Add all artifacts to list to start.
    $.each(state.artifacts, function(artifact_name, value) {
      grr.artifact_view.artifact_manager.add(artifact_div, artifact_name);
    });

    // Add all OS options to os filter.
    state.supported_os.push('All');
    $.each(state.supported_os, function(artifact_name, value) {
      grr.artifact_view.list_manager.add(os_filter_element, value);
    });
    grr.artifact_view.updateFilter(artifact_div);

    // Search Handler.
    $('#' + state.unique + '_search').keyup(function() {
      grr.artifact_view.updateFilter(artifact_div);
    });

    $('#' + os_filter_element).change(function() {
      grr.artifact_view.updateFilter(artifact_div);
    });

    // Artifact Handlers

    // Add on doubleclick.
    $('#' + artifact_list_element).dblclick(function() {
      grr.artifact_view.artifact_manager.add_selected(artifact_div,
          [this.value]);
      grr.forms.selectOnChange('#' + selected_list_element);
    });

    // Add all.
    $('#' + state.unique + '_artifact_add_all').click(function() {
      var all_opts = $('#' + artifact_list_element + ' option');
      var all_opts_vals = [];
      all_opts.each(function(index, value) {
        all_opts_vals.push(value.value);
      });
      grr.artifact_view.artifact_manager.add_selected(artifact_div,
          all_opts_vals);
      grr.forms.selectOnChange('#' + selected_list_element);
    });

    // Add.
    $('#' + state.unique + '_artifact_add').click(function() {
      var selected = $('#' + artifact_list_element).val();
      grr.artifact_view.artifact_manager.add_selected(artifact_div, selected);
      grr.forms.selectOnChange('#' + selected_list_element);
    });

    // Render Artifact information on select.

    $('#' + artifact_div.data('description_element')).hide(0);
    var select_elements = $('#' + artifact_list_element + ',#' +
        selected_list_element);
    select_elements.on('change select', function() {
      grr.artifact_view.renderArtifactFromDom(artifact_div, this.value,
          description_element);
    });

    // Select List Handlers

    // Remove on doubleclick.
    $('#' + selected_list_element).dblclick(function() {
      grr.artifact_view.artifact_manager.remove_from_selection(artifact_div,
          this.value);
    });

    // Remove.
    $('#' + state.unique + '_select_remove').click(function() {
      var selected = $('#' + selected_list_element).val();
      grr.artifact_view.artifact_manager.remove_from_selection(artifact_div,
          selected);
      grr.forms.selectOnChange('#' + selected_list_element);
    });

    // Clear.
    $('#' + state.unique + '_select_clear').click(function() {
      grr.artifact_view.artifact_manager.clear(artifact_div);
      grr.forms.selectOnChange('#' + selected_list_element);
    });
  }
});



/**
 * Update list of artifacts based on filter options.
 *
 * @param {object} artifact_div The div containing the artifact renderer.
 *
 */
grr.artifact_view.updateFilter = function(artifact_div) {
  var search_string = $('#' + artifact_div.data('search_element')).val();
  var os_filter = $('#' + artifact_div.data('os_filter_element')).val();
  grr.artifact_view.artifact_manager.filter(artifact_div, search_string,
      os_filter);
};


/**
 * Renders an artifact into an element from an initialized artifact dom.
 *
 * @param {object} artifact_div The div containing the artifact renderer.
 * @param {string} artifact_name name of artifact
 * @param {string} element to write artifact into.
 */
grr.artifact_view.renderArtifactFromDom = function(artifact_div, artifact_name,
    element) {
  artifact = artifact_div.data('artifacts')[artifact_name];
  if (! artifact) {
    return;
  }
  $('#' + artifact_div.data('description_element')).show(0);
  grr.artifact_view.renderArtifactFromObject(artifact, element);
};


/**
 * Renders an artifact into an element.
 *
 * @param {object} artifact An object containing the artifact.
 * @param {string} element to write artifact into.
 */
grr.artifact_view.renderArtifactFromObject = function(artifact, element) {
  $('#' + element + ' div[name=artifact_name]').text(artifact.name);
  $('#' + element + ' div[name=artifact_labels]').text(artifact.labels);
  var desc_element = $('#' + element + ' div[name=artifact_description]');
  // Set text from description, but allow for newlines.
  var description = desc_element.text(artifact.doc).html();
  description = description.replace(/\n/g, '<br />');
  desc_element.html(description);

  $('#' + element +
      ' div[name=artifact_conditions]').text(artifact.conditions);
  $('#' + element +
      ' div[name=artifact_dependencies]').text(artifact.dependencies);
  $('#' + element +
      ' div[name=artifact_supported_os]').text(artifact.supported_os);
  $('#' + element +
      ' div[name=artifact_output_type]').text(artifact.output_type);

  var links_element = $('#' + element + ' div[name=artifact_links]');
  links_element.html('');
  if (artifact.urls.length > 0) {
    $.each(artifact.urls, function(index, link) {
      var link_html = ('<a href="' + link + '" "noreferrer" target="_blank">' +
          link + '</a><br/>');
      links_element.append(link_html);
    });
  }

  var processor_element = '#' + element + ' table[name=artifact_processors]';
  $(processor_element + ' tr').remove();
  if (artifact.processors && artifact.processors.length > 0) {
    $.each(artifact.processors, function(index, processor) {
      processor_row = '<tr><td>Parser<td>' + processor.name + '</tr>';
      processor_row += '<tr><td>Output types<td>' + processor.output_types +
          '</tr>';
      processor_row += '<tr><td>Description<td>' + processor.doc +
          '</tr>';
      processor_row += '<tr><td></tr>';
      $(processor_element).append(processor_row);
    });
  } else {
    $(processor_element).append('<tr><td>None</td></tr>');
  }

  var source_element = '#' + element + ' table[name=artifact_sources]';
  $(source_element + ' tr').remove();
  if (artifact.sources.length > 0) {
    $.each(artifact.sources, function(index, source) {
      source_row = '<tr><td>Type<td>' + source.type +
        '</tr>';
      $.each(source.attributes, function(name, value) {
        if ($.isArray(value)) {
          value = value.join('<br>');
        }
        source_row += '<tr><td>arg:' + name + '<td>' + value + '</tr>';
      });
      source_row += '<tr><td></tr>';
      $(source_element).append(source_row);
    });
  } else {
    $(source_element).append('<tr><td>None</tr>');
  }

  $('#' + element + ' td:first-child').addClass('proto_key');
  $('#' + element + ' td:nth-child(2)').addClass('proto_value');
};


/**
 * Select List Manager object for handling a multi-select box.
 */
grr.artifact_view.list_manager = {};

/**
 * Add an element to the list.
 *
 * @param {string} domId The id of the select element to add to.
 * @param {string} value The primary value for the object.
 * @param {object} value_object Object to associate with the value.
 *
 */
grr.artifact_view.list_manager.add = function(domId, value, value_object) {
  var data = $('#' + domId).data();

  if (!(value in data)) {
    var row = '<option value="' + value + '">' + value + '</option>';
    $('#' + domId).append(row);
    data[value] = value_object;
  }
};

/**
 * Remove an element from the list.
 *
 * @param {string} domId The id of the select element to remove from.
 * @param {string} value The primary value for the object to remove.
 *
 */
grr.artifact_view.list_manager.remove = function(domId, value) {
  if (value) {
    $('#' + domId + ' option[value=' + value + ']').remove();
    $('#' + domId).removeData(value);
  }
};

/**
 * Clear the list.
 *
 * @param {string} domId The id of the select element to update to.
 *
 */
grr.artifact_view.list_manager.clear = function(domId) {
  $('#' + domId).html('');
  $('#' + domId).removeData();
};



/**
 * Artifact Manager helper object and functions.
 */
grr.artifact_view.artifact_manager = {};

/**
 * Add an element to the artifact list.
 *
 * @param {object} artifact_div The div containing the artifact renderer.
 * @param {string} artifact_name The name of the artifact.
 *
 */
grr.artifact_view.artifact_manager.add = function(artifact_div, artifact_name) {
  var artifact_list = artifact_div.data('artifact_list_element');
  grr.artifact_view.list_manager.add(artifact_list, artifact_name,
      [artifact_div.data('artifacts')[artifact_name].description]);
};

/**
 * Add artifacts to the selected list.
 *
 * @param {object} artifact_div The div containing the artifact renderer.
 * @param {string} artifact_names The names of the artifacts to add.
 *
 */
grr.artifact_view.artifact_manager.add_selected = function(artifact_div,
    artifact_names) {
  $.each(artifact_names, function(index, artifact_name) {
    var artifact = artifact_div.data('artifacts')[artifact_name];
    if (artifact) {
      grr.artifact_view.list_manager.add(
          artifact_div.data('selected_list_element'),
          artifact_name, [artifact.doc]);
    }
  });
};

/**
 * Remove artifact from the selected list.
 *
 * @param {object} artifact_div The div containing the artifact renderer.
 * @param {string} artifact_name The name of the artifact to remove.
 *
 */
grr.artifact_view.artifact_manager.remove_from_selection = function(
    artifact_div, artifact_name) {
  grr.artifact_view.list_manager.remove(
      artifact_div.data('selected_list_element'), artifact_name);
};

/**
 * Clear the selected artifact list.
 *
 * @param {object} artifact_div The div containing the artifact renderer.
 *
 */
grr.artifact_view.artifact_manager.clear = function(artifact_div) {
  grr.artifact_view.list_manager.clear(
      artifact_div.data('selected_list_element'));
};

/**
 * Filter artifact list by search string.
 *
 * @param {object} artifact_div The div containing the artifact renderer.
 * @param {string} search_string String to search artifact properties for.
 * @param {string} os The OS to limit to, if Any, do no filtering.
 *
*/
grr.artifact_view.artifact_manager.filter = function(artifact_div,
    search_string, os) {
  var artifact_list = artifact_div.data('artifact_list_element');
  grr.artifact_view.list_manager.clear(artifact_list);
  var artifact_temp = [];
  var search_string = search_string.toLowerCase();
  $.each(artifact_div.data('artifacts'),
      function(artifact_name, artifact) {
    // Filter based on supported_os.
    if ((artifact.supported_os.length == 0) ||
        ($.inArray(os, artifact.supported_os) >= 0)) {
      // Filter based on search string.
      if (artifact.doc.toLowerCase().search(search_string) != -1 ||
          artifact_name.toLowerCase().search(search_string) != -1) {
        artifact_temp.push(artifact_name);
      }
      $.each(artifact.labels, function(index, label_name) {
        if (label_name.toLowerCase().search(search_string) != -1) {
          artifact_temp.push(artifact_name);
          return false;
        }
      });
    }
  });
  // Now sort and add them.
  artifact_temp.sort();
  $.each(artifact_temp, function(index, artifact_name) {
    grr.artifact_view.artifact_manager.add(artifact_div, artifact_name);
  });

};
